--- a\srtcore\common.cpp
+++ b\srtcore\common.cpp
@@ -367,7 +367,7 @@
 {
    md5_state_t state;
 
-   md5_init(&state);
+   md5_init_srt(&state);
    md5_append(&state, (const md5_byte_t *)input, (int) strlen(input));
    md5_finish(&state, result);
 }
--- a\srtcore\md5.h
+++ b\srtcore\md5.h
@@ -76,7 +76,7 @@
 #endif
 
 /* Initialize the algorithm. */
-void md5_init(md5_state_t *pms);
+void md5_init_srt(md5_state_t *pms);
 
 /* Append a string to the message. */
 void md5_append(md5_state_t *pms, const md5_byte_t *data, int nbytes);
--- a\srtcore\md5.cpp
+++ b\srtcore\md5.cpp
@@ -310,7 +310,7 @@
 }
 
 void
-md5_init(md5_state_t *pms)
+md5_init_srt(md5_state_t *pms)
 {
     pms->count[0] = pms->count[1] = 0;
     pms->abcd[0] = 0x67452301;
